### PR TITLE
feat: withdrawal handler

### PR DESCRIPTION
**Pull request by ermvrs**
https://github.com/keep-starknet-strange/satoru/issues/309
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing
- Other (please describe):

## What is the current behavior?

Resolves: #NA

## What is the new behavior?

-
-
-

### CODE REVIEW

**Comment by ermvrs** [comment]
I know there is some unnecessary comments. Please let me know if everything right with this implementation cause I changed some other things aswell. After you guys confirm ill clear comments.

File: src/exchange/withdrawal_handler.cairo [hunk 1]
```diff
@@ -66,13 +54,18 @@ mod WithdrawalHandler {
...
-    use starknet::ContractAddress;
-
+    use starknet::{ContractAddress, get_contract_address, get_caller_address};
+    use traits::Default;
+    use clone::Clone;
 
     // Local imports.
-    use super::IWithdrawalHandler;
+    use super::{
+        IWithdrawalHandler, IWithdrawalHandlerDispatcher, IWithdrawalHandlerDispatcherTrait
```
Comment (zarboq) [inline src/exchange/withdrawal_handler.cairo:63]: You can remove the dispatcher import

File: src/exchange/withdrawal_handler.cairo [hunk 2]
```diff
@@ -182,7 +308,60 @@ mod WithdrawalHandler {
...
+                reason_bytes
+            );
+        }
+
+        /// Executes a withdrawal with keeper.
+        /// # Arguments
+        /// * `key` - The key of the withdrawal to execute.
+        /// * `oracle_params` - The oracle params to set prices before execution.
+        /// * `keeper` - The keeper executing the withdrawal.
+        fn execute_withdrawal_keeper(
```
Comment (zarboq) [inline src/exchange/withdrawal_handler.cairo:331]: This is marked as external in GMX code

File: src/exchange/withdrawal_handler.cairo [hunk 3]
```diff
@@ -66,13 +54,18 @@ mod WithdrawalHandler {
...
+    use traits::Default;
+    use clone::Clone;
 
     // Local imports.
-    use super::IWithdrawalHandler;
+    use super::{
+        IWithdrawalHandler, IWithdrawalHandlerDispatcher, IWithdrawalHandlerDispatcherTrait
+    };
     use satoru::role::role_store::{IRoleStoreDispatcher, IRoleStoreDispatcherTrait};
+    use satoru::role::role;
```
Comment (zarboq) [inline src/exchange/withdrawal_handler.cairo:66]: refactor imports

```
use satoru::role::role_store::{IRoleStoreDispatcher, IRoleStoreDispatcherTrait};
use satoru::role::role;
```
becomes:
```
use satoru::role::{role, role_store::{IRoleStoreDispatcher, IRoleStoreDispatcherTrait};
```

Do it for other imports also please

**Review by zarboq — CHANGES_REQUESTED** [review]

File: src/exchange/withdrawal_handler.cairo (hunk 2 above)
Comment (ermvrs) [reply src/exchange/withdrawal_handler.cairo:331]: I don't think there is a way to call contract itself with a msg.sender context of this contract. In GMX there is a check msg.sender == address(this) so as I know it's not possible to achieve this currently in Starknet. so I moved this function to internal

Comment (zarboq) [reply src/exchange/withdrawal_handler.cairo:331]: Yeah you're right didn't see that !

**Review by zarboq — APPROVED** [review]

**Pull request merged**